docs/fsharp/language-reference/results.md

```diff
@@ -0,0 +1,94 @@
+---
+title: Results (F#)
+description: Results (F#)
+keywords: visual f#, f#, functional programming
+author: cartermp
+ms.author: phcart
+ms.date: 04/24/2017
+ms.topic: language-reference
+ms.prod: .net
+ms.technology: devlang-fsharp
+ms.devlang: fsharp
+ms.assetid: a15b5cf1-9055-4481-918c-4c8a051b5829
+---
+
+# Results
+
+Starting with F# 4.1, there is a `Result<'T,'TFailure>` type which you can use for writing error-tolerant code which can be composed monadically.
+
+## Syntax
+
+```fsharp
+// The definition of Result in FSharp.Core
+[<StructuralEquality; StructuralComparison>]
+[<CompiledName("FSharpResult`2")>]
+[<Struct>]
+type Result<'T,'TError> = 
+    | Ok of ResultValue:'T 
+    | Error of ErrorValue:'TError
+```
+
+## Remarks
+
+Note that the result type is a [struct discriminated union](discriminated-unions.md#struct-discriminated-unions), which is another feature introduced in F# 4.1.  Structural equality semantics apply here.
+
+The `Result` type is typically used in monadic error-handling, which is often referred to as [Railway-oriented Programming](https://swlaschin.gitbooks.io/fsharpforfunandprofit/content/posts/recipe-part2.html) within the F# community.  The following trivial example demonstrates this approach.
+
+```fsharp
+// Define a simple type which has fields that can be validated
+type Request = {
+    Name: string
+    Email: string
+}
+
+// Define an infix operator for Result.bind to compose
+let (<<=) input f = bind f input
```

Comment:
There's another glitch here - you define ``<<=`` but use ``>>=``.  But I actually think it's better to not define any new operator, and simply use ``|> Refult.bind validateEmail |> Result.bind validateName``. See full working code below.  
```fsharp
open System

type Request = 
    { Name: string
      Email: string }

// Define a simple type which has fields that can be validated
let validateName req =
    match req.Name with
    | null -> Error "No name found."
    | "" -> Error "Name is empty."
    | "bananas" -> Error "Bananas is not a name."
    | _ -> Ok req

// Similarily, define some email validation logic.
let validateEmail req =
    match req.Email with
    | null -> Error "No email found."
    | "" -> Error "Email is empty."
    | s when s.EndsWith("bananas.com") -> Error "No email from bananas.com is allowed."
    | _ -> Ok req

// Compose the validation functions
let validateRequest req =
    req 
    |> Result.bind validateName
    |> Result.bind validateEmail 

let test() = 

    // Now, create a Request and pattern match on the result.
    let req1 = { Name = "Phillip"; Email = "phillip@contoso.biz" }
    let res1 = validateRequest (Ok req1)

    match res1 with
    | Ok req -> printfn "My request was valid! Name: %s Email %s" req1.Name req1.Email
    | Error e -> printfn "Error: %s" e

    // Prints " "My request was valid!  Name: Phillip Email: phillip@consoto.biz"

    let req2 = { Name = "Phillip"; Email = "phillip@bananas.com" }
    let res2 = validateRequest  (Ok req2)

    match res2 with
    | Ok req -> printfn "My request was valid! Name: %s Email %s" req2.Name req2.Email
    | Error e -> printfn "Error: %s" e

    // Prints: "Error: No email from bananas.com is allowed."

test()
```